BraintreeCore/src/main/java/com/braintreepayments/api/PostalAddressParser.kt

```diff
@@ -0,0 +1,89 @@
+    fun fromJson(accountAddress: JSONObject?): PostalAddress =
+        // If we don't have an account address, return an empty PostalAddress.
+        accountAddress?.let {
+            var streetAddress = Json.optString(accountAddress, STREET_ADDRESS_KEY, null)
+            var extendedAddress = Json.optString(accountAddress, EXTENDED_ADDRESS_KEY, null)
+            var countryCodeAlpha2 = Json.optString(accountAddress, COUNTRY_CODE_ALPHA_2_KEY, null)
+
+            //Check alternate keys
+            streetAddress = streetAddress ?: Json.optString(accountAddress, LINE_1_KEY, null)
+            extendedAddress = extendedAddress ?: Json.optString(accountAddress, LINE_2_KEY, null)
+            countryCodeAlpha2 = countryCodeAlpha2 ?: Json.optString(accountAddress, COUNTRY_CODE_KEY, null)
+
+            // If this is a UserAddress-like JSON, parse it as such
+            if (streetAddress == null && Json.optString(
+                    accountAddress,
+                    USER_ADDRESS_NAME_KEY,
+                    null
+                ) != null
+            ) {
+                return@let fromUserAddressJson(accountAddress)
+            }
+
+            PostalAddress().apply {
+                recipientName = Json.optString(accountAddress, RECIPIENT_NAME_KEY, null)
+                this.streetAddress = streetAddress
+                this.extendedAddress = extendedAddress
+                locality = Json.optString(accountAddress, LOCALITY_KEY, null)
+                region = Json.optString(accountAddress, REGION_KEY, null)
+                postalCode = Json.optString(accountAddress, POSTAL_CODE_KEY, null)
+                this.countryCodeAlpha2 = countryCodeAlpha2
+            }
+
+        } ?: PostalAddress()
+
+    fun fromUserAddressJson(json: JSONObject): PostalAddress =
+        PostalAddress().apply {
+            recipientName = Json.optString(json, USER_ADDRESS_NAME_KEY, "")
+            phoneNumber = Json.optString(json, USER_ADDRESS_PHONE_NUMBER_KEY, "")
+            streetAddress = Json.optString(json, USER_ADDRESS_ADDRESS_1_KEY, "")
+            extendedAddress = formatExtendedUserAddress(json)
+            locality = Json.optString(json, USER_ADDRESS_LOCALITY_KEY, "")
+            region = Json.optString(json, USER_ADDRESS_ADMINISTRATIVE_AREA_KEY, "")
+            countryCodeAlpha2 = Json.optString(json, USER_ADDRESS_COUNTRY_CODE_KEY, "")
+            postalCode = Json.optString(json, USER_ADDRESS_POSTAL_CODE_KEY, "")
+            sortingCode = Json.optString(json, USER_ADDRESS_SORTING_CODE_KEY, "")
+        }
+
+    private fun formatExtendedUserAddress(address: JSONObject): String =
+        """
+            ${Json.optString(address, USER_ADDRESS_ADDRESS_2_KEY, "")}
+            ${Json.optString(address, USER_ADDRESS_ADDRESS_3_KEY, "")}
+            ${Json.optString(address, USER_ADDRESS_ADDRESS_4_KEY, "")}
+            ${Json.optString(address, USER_ADDRESS_ADDRESS_5_KEY, "")}
+        """.trimIndent().trim { it <= ' ' }
```

Comment:
Does something like this work in Kotlin? The `.trimIndent().trim { it <= ' ' }` is a little tricky to read

```suggestion
        return
            "${Json.optString(address, USER_ADDRESS_ADDRESS_2_KEY, "")} + "\n" +
            "${Json.optString(address, USER_ADDRESS_ADDRESS_3_KEY, "")} + "\n" +
            "${Json.optString(address, USER_ADDRESS_ADDRESS_4_KEY, "")} + "\n" +
            "${Json.optString(address, USER_ADDRESS_ADDRESS_5_KEY, "")}"
```

Reply:
Yeah is `<=` a Kotlin operator?

Reply:
Oh wow definitely less than or equal too 🤦 . This looks like it's trimming characters that have a numeric value of at lease the space character. We probably don't need the additional trimming.